rslib/src/cloze.rs

```diff
@@ -1,129 +1,294 @@
+                    }
+                    cloze.nodes.push(TextOrCloze::Text(text));
                 } else {
-                    "..."
+                    output.push(TextOrCloze::Text(text));
                 }
+            }
+            Token::CloseCloze => {
+                // take the currently active cloze
+                if let Some(cloze) = open_clozes.pop() {
+                    let target = if let Some(outer_cloze) = open_clozes.last_mut() {
+                        // and place it into the cloze layer above
+                        &mut outer_cloze.nodes
+                    } else {
+                        // or the top level if no other clozes active
+                        &mut output
+                    };
+                    target.push(TextOrCloze::Cloze(cloze));
+                } else {
+                    // closing marker outside of any clozes
+                    output.push(TextOrCloze::Text("}}"))
+                }
+            }
+        }
+    }
+    output
+}
+
+fn reveal_cloze_text_in_nodes(
+    node: &TextOrCloze,
+    cloze_ord: u16,
+    question: bool,
+    output: &mut Vec<String>,
+) {
+    if let TextOrCloze::Cloze(cloze) = node {
+        if cloze.ordinal == cloze_ord {
+            if question {
+                output.push(cloze.hint().into())
             } else {
-                caps.get(cloze_caps::TEXT).unwrap().as_str()
-            };
+                output.push(cloze.clozed_text().into())
+            }
+        }
+        for node in &cloze.nodes {
+            reveal_cloze_text_in_nodes(node, cloze_ord, question, output);
+        }
+    }
+}
 
-            cloze
-        })
-        .collect::<Vec<_>>()
-        .join(", ")
-        .into()
+fn reveal_cloze(
+    cloze: &ExtractedCloze,
+    cloze_ord: u16,
+    question: bool,
+    active_cloze_found_in_text: &mut bool,
+    buf: &mut String,
+) {
+    let active = cloze.ordinal == cloze_ord;
+    *active_cloze_found_in_text |= active;
+    match (question, active) {
+        (true, true) => {
+            // question side with active cloze; all inner content is elided
+            let mut content_buf = String::new();
+            for node in &cloze.nodes {
+                match node {
+                    TextOrCloze::Text(text) => content_buf.push_str(text),
+                    TextOrCloze::Cloze(cloze) => reveal_cloze(
+                        cloze,
+                        cloze_ord,
+                        question,
+                        active_cloze_found_in_text,
+                        &mut content_buf,
+                    ),
+                }
+            }
+            buf.push_str(&format!(
+                r#"<span class="cloze" data-cloze="{}" data-ordinal="{}">[{}]</span>"#,
+                encode_attribute(&content_buf),
+                cloze.ordinal,
+                cloze.hint()
+            ));
+        }
+        (false, true) => {
+            buf.push_str(&format!(
```

Comment:
It's a little more efficient here to do

```
writeln!(buf, "...", ...).unwrap()
```

Same with other lines which call `push_str(&format...)`

Reply:
Ok, updated.

Reply:
Hmm, seems like I missed that and just saw the rebuild triggered from update merge from main.